Since Vaadin 14.0.9 and 14.0.10 production mode not available

Since Vaadin **14.0.9** and **14.0.10** the problem **occurs ** that -Pproduction ist not generating a production package. Vaadin is still running in debug mode.

I am using Spring Boot: 2.2.0, Java JDK 13.0.1 and NPM 12.X.

You can reproduce the problem by using my artefact. The source code and build is documented on GitHub. Please see my links:

- [Sourcecode on GitHub](https://github.com/christian-ehmke/vaadin-starter-project)
- [Build on GitHub](https://github.com/christian-ehmke/vaadin-starter-project/commit/24b6c4948197c6c253f49bdf7cbf53c3cc9d39c0/checks?check_suite_id=277367460)
- [Download JAR fromGitHub](https://github.com/christian-ehmke/vaadin-starter-project/suites/277367460/artifacts/163164)

The problem occurs on my local Windows machine, Google Cloud Compute Engine (Ubuntu 19.04) and inside a Docker Container (Ubuntu 19.04) running on Google Cloud Run.

In the console you will see:
```
Vaadin is running in DEBUG MODE.
Add productionMode=true to web.xml to disable debug features.
```
In the browser you can see also:
```
Webpack Error
ERROR in ../target/frontend/generated-flow-imports.js
Module not found: Error: Can't resolve '@vaadin/flow-frontend/flow-component-renderer.js' in '/home/admin_christian_ehmke_de/work/my-starter-project-2.0-SNAPSHOT.jar/target/frontend'
 @ ../target/frontend/generated-flow-imports.js 12:0-58

ERROR in ../target/frontend/generated-flow-imports.js
Module not found: Error: Can't resolve '@vaadin/flow-frontend/flow-component-renderer.js' in '/home/admin_christian_ehmke_de/work/my-starter-project-2.0-SNAPSHOT.jar/target/frontend'
 @ ../target/frontend/generated-flow-imports.js 12:0-58
ℹ ｢wdm｣: Failed to compile.
```

Switching back to Vaadin 14.0.8 solves the problem.

_Originally posted by @christian-ehmke in https://github.com/vaadin/flow/issues/6677#issuecomment-545414378_

I have updated to Vaadin 14.0.11 but the problem still occurs:

- [Sourcecode on GitHub](https://github.com/christian-ehmke/vaadin-starter-project)
- [Build on GitHub](https://github.com/christian-ehmke/vaadin-starter-project/commit/c3bd1782a7f6f98b9e7c7d665daae287b0d15e28/checks?check_suite_id=286153566)
- [Download JAR fromGitHub](https://github.com/christian-ehmke/vaadin-starter-project/suites/286153566/artifacts/200808)

```
====================================================================
Vaadin is running in DEBUG MODE.
In order to run your application in production mode and disable debug features, you should enable it by setting the servlet init parameter productionMode to true.
See https://vaadin.com/docs/v14/flow/production/tutorial-production-mode-basic.html for more information about the production mode.
====================================================================
```

```
2019-10-29 12:45:07.821  INFO 2323 --- [        webpack] dev-webpack                              :     + 173 hidden modules
2019-10-29 12:45:07.821  INFO 2323 --- [        webpack] dev-webpack                              : 
2019-10-29 12:45:07.821 ERROR 2323 --- [        webpack] dev-webpack                              : ERROR in ../target/frontend/generated-flow-imports.js
2019-10-29 12:45:07.822 ERROR 2323 --- [        webpack] dev-webpack                              : Module not found: Error: Can't resolve '@vaadin/flow-frontend/flow-component-renderer.js' in '/home/admin_christian_ehmke_de/my-starter-project-2.0-SNAPSHOT.jar/target/frontend'
2019-10-29 12:45:07.822 ERROR 2323 --- [        webpack] dev-webpack                              :  @ ../target/frontend/generated-flow-imports.js 12:0-58
2019-10-29 12:45:07.822 ERROR 2323 --- [        webpack] dev-webpack                              : 
2019-10-29 12:45:07.822 ERROR 2323 --- [        webpack] dev-webpack                              : ERROR in ../target/frontend/generated-flow-imports.js
2019-10-29 12:45:07.824 ERROR 2323 --- [        webpack] dev-webpack                              : Module not found: Error: Can't resolve '@vaadin/flow-frontend/flow-component-renderer.js' in '/home/admin_christian_ehmke_de/my-starter-project-2.0-SNAPSHOT.jar/target/frontend'
2019-10-29 12:45:07.824 ERROR 2323 --- [        webpack] dev-webpack                              :  @ ../target/frontend/generated-flow-imports.js 12:0-58
2019-10-29 12:45:07.824 ERROR 2323 --- [        webpack] dev-webpack                              : ℹ ｢wdm｣: Failed to compile.
2019-10-29 12:45:07.824 ERROR 2323 --- [        webpack] dev-webpack                              : 
------------------ Frontend compilation failed. -----------------
2019-10-29 12:45:08.339  INFO 2323 --- [        webpack] dev-webpack                              : Watchdog connection closed. Trying to re-run watchdog.
```

--- a/flow-server/src/main/java/com/vaadin/flow/server/DeploymentConfigurationFactory.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/DeploymentConfigurationFactory.java
@@ -242,7 +242,7 @@ public final class DeploymentConfigurationFactory implements Serializable {
         try {
             json = getResourceFromFile(initParameters);
             if (json == null) {
-                json = getResourceFromClassloader(initParameters);
+                json = getResourceFromClassloader();
             }
         } catch (IOException e) {
             throw new UncheckedIOException(e);
@@ -265,7 +265,7 @@ public final class DeploymentConfigurationFactory implements Serializable {
         return json;
     }
 
-    private static String getResourceFromClassloader(Properties initParameters)
+    private static String getResourceFromClassloader()
             throws IOException {
         String json = null;
         // token file is in the class-path of the application
@@ -278,10 +278,11 @@ public final class DeploymentConfigurationFactory implements Serializable {
         URL resource = resources.stream()
                 .filter(url -> !url.getPath().endsWith("jar!/" + tokenResource))
                 .findFirst().orElse(null);
-        if (resource == null && isProductionMode(initParameters)) {
-            // For no non jar build info, in production mode check jar files
-            // for production mode jar.
-            json = getProductionModeResource(resources);
+        if (resource == null && !resources.isEmpty()) {
+            // For no non jar build info, in production mode check for
+            // webpack.generated.json if it's in a jar in a jar then accept
+            // single jar flow-build-info.
+            json = getPossibleJarResource(resources);
         } else if (resource != null) {
             json = FrontendUtils.streamToString(resource.openStream());
         }
@@ -315,24 +316,29 @@ public final class DeploymentConfigurationFactory implements Serializable {
     }
 
     /**
-     * Check resources for a production mode resource if all resources were
-     * from JAR files as we may be running from a JAR and add-ons are not
-     * packaged in productionMode.
+     * Check if the webpack.generated.js resources is inside 2 jars
+     * (flow-server.jar and application.jar) if this is the case then we can
+     * accept a build info file from inside  jar with a single jar in the path.
      *
      * @param resources
      *         flow-build-info url resource files
-     * @return production mode flow-build-info string
+     * @return flow-build-info json string or <code>null</code> if no applicable files found
      * @throws IOException
      *         exception reading stream
      */
-    private static String getProductionModeResource(List<URL> resources)
+    private static String getPossibleJarResource(List<URL> resources)
             throws IOException {
-        for (URL resource : resources) {
-            String json = FrontendUtils.streamToString(resource.openStream());
-            if (json != null && !json.contains(NPM_TOKEN) && !json
-                    .contains(FRONTEND_TOKEN) && json
-                    .contains("\"productionMode\": true")) {
-                return json;
+        URL webpackGenerated = DeploymentConfiguration.class.getClassLoader()
+                .getResource(FrontendUtils.WEBPACK_GENERATED);
+        // If jar!/ exists 2 times for webpack.generated.json then we are
+        // running from a jar
+        if (webpackGenerated.getPath().split("jar!/", -1).length - 1 >= 2) {
+            for (URL resource : resources) {
+                // As we are running from a war we can accept a info with a
+                // single jar in the path
+                if (resource.getPath().split("jar!/", -1).length - 1 == 1) {
+                    return FrontendUtils.streamToString(resource.openStream());
+                }
             }
         }
         // No applicable resources found.

--- a/flow-server/src/main/java/com/vaadin/flow/server/DeploymentConfigurationFactory.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/DeploymentConfigurationFactory.java
@@ -332,17 +332,21 @@ public final class DeploymentConfigurationFactory implements Serializable {
                 .getResource(FrontendUtils.WEBPACK_GENERATED);
         // If jar!/ exists 2 times for webpack.generated.json then we are
         // running from a jar
-        if (webpackGenerated.getPath().split("jar!/", -1).length - 1 >= 2) {
+        if (countInstances(webpackGenerated.getPath(), "jar!/") >= 2) {
             for (URL resource : resources) {
-                // As we are running from a war we can accept a info with a
-                // single jar in the path
-                if (resource.getPath().split("jar!/", -1).length - 1 == 1) {
+                // As we now know that we are running from a jar we can accept a
+                // build info with a single jar in the path
+                if (countInstances(resource.getPath(), "jar!/") == 1) {
                     return FrontendUtils.streamToString(resource.openStream());
                 }
             }
         }
         // No applicable resources found.
         return null;
+    }
+
+    private static int countInstances(String input, String value) {
+        return input.split(value, -1).length - 1;
     }
 
     /**

